enhancements/machine-api/machine-deletion-hooks.md

```diff
@@ -0,0 +1,471 @@
+has not progressed in removing the Machine. It is expected that hook implementing
+controllers should signal to users when they are having issues via some mechanism
+(eg. their `ClusterOperator` status) and that the conditions on the Machine should
+identify the components which the user should look at in order to determine any
+underlying issues.
+
+For example, the condition will contain the owner of the hook blocking the removal,
+the user should look towards the owner's own reporting to identify issues.
+Any OpenShift operator experiencing issues should already have a mechanism in which
+it can report errors and having these errors follow the normal mechanisms should
+place the failure reporting closer to the issue, as opposed to placing it directly
+onto the Machine.
+
+We also encourage hook implementing controllers to use Event objects to add detail
+about the operations they are performing. For example if the hook implementing
+controller is waiting on some side effect, it should periodically send an Event,
+attached to the Machine object, notifying the user of delay. Events are already
+displayed in the OpenShift console and in `oc describe` and as such should be
+relatively easy for the end user to find.
+
+##### Hook ordering
+
+The machine controller will not attempt to enforce any ordering of hooks.
+No ordering should be expected by the machine controller.
+
+Hook Implementing Controllers may choose to provide a mechanism to allow
+ordering amongst themselves via whatever means HICs determine.
+Examples could be using CRDs external to the Machine API, gRPC communications,
+or additional annotations on the Machine or other objects.
+
+#### Hook Implementing Controller Design
+
+Hook Implementing Controller is the component that manages a particular
+lifecycle hook.
+
+##### Hook Implementing Controllers must
+
+* Watch Machine objects and determine when an appropriate action must be taken.
+* After completing the desired hook action, remove the hook annotation.
+
+##### Hook Implementing Controllers may
+
+* Watch Machine objects and add a hook annotation as desired by the cluster
+administrator.
+* Coordinate with other Hook Implementing Controllers through any means
+possible, such as using common annotations, CRDs, etc. For example, one hook
+controller could set an annotation indicating it has finished its work, and
+another hook controller could wait for the presence of the annotation before
+proceeding.
+
+#### Determining when to take action
+
+A Hook Implementing Controller should watch Machines and determine when is the
+best time to take action.
+
+For example, if an HIC manages a lifecycle hook at the pre-drain lifecycle-point,
+then that controller should take action immediately after a Machine has a
+DeletionTimestamp or enters the "Deleting" phase.
+
+Fine-tuned coordination is not possible at this time; eg, it's not
+possible to execute a pre-terminate hook only after a node has been drained.
```

Comment:
This isn't obvious to me.  Why can a hook implementing controller not detect that the node has been drained?